GVFS/GVFS.UnitTests/Mock/Git/MockGitProcess.cs

```diff
@@ -17,17 +15,33 @@ public MockGitProcess()
             : base(new MockGVFSEnlistment())
         {
             this.CommandsRun = new List<string>();
+            this.StoredCredentials = new Dictionary<string, IList<Credential>>(StringComparer.OrdinalIgnoreCase);
         }
 
         public List<string> CommandsRun { get; }
         public bool ShouldFail { get; set; }
+        public IDictionary<string, IList<Credential>> StoredCredentials { get; }

```

Comment:
Is there a reason why this is `IDictionary` vs `Dictionary`?

Reply:
My default is to use `IFoo` rather than `Foo` when the member is `public`, even if that's only public to others in the assembly. No one else cares what the exact implementation is, and they don't need to use any members only exposed on `Foo` but not `IFoo`. If they do they can change it.

---

Aside:

I've noticed there seems to be a huge aversion in the team towards using interfaces or other abstract types, out of some fear I don't fully understand or buy\*.

Is "be liberal in what you accept and conservative in what you return" (even within in the same codebase) not a widely accepted ethos?

\* I know there is some opposition when it comes to the LINQ-based collection methods vs the concrete-types' O(1) properties, which is a valid concern when it comes to perf which I _do_ buy. However my counterargument to avoiding the `ICollection` etc interfaces is that:

1. Treating developers as 'dangerous' and introducing arbitrary rule/restrictions/commandments on usage or non-usage of language and framework constructs & patterns is silly. (It's why I'm also an opponent of some of the more pedantic StyleCop rules around comments and whitespace.. it's just more frustrating than anything for the perceived benefits.)
Each usage or non-usage should be taken in it's own context as to if it's appropriate, or not, or if it doesn't really matter.
2. Many of the LINQ calls actually have [fast paths](https://github.com/dotnet/corefx/blob/53ae9af48c5c133f4413e76c1469dfe6ccecd926/src/System.Linq/src/System/Linq/Count.cs#L19-L22) for the LINQ calls, for example `new List().Count()` actually calls through to the `.Count` property. Not that I'm advocating for using `.Count()` over `.Count` but any accidental usage isn't actually that impactful. 
3. Any good IDE, static analyser, and code review would notice "usage of `.Count()` when `.Count` is available". In fact if a code review misses `.Count()` vs `.Count` then it's also possible the same review would miss `IList` vs `List` too.

Reply:
Another point I'd like to bring up is consistency.  StyleCop helps quite a bit in this area, but it cannot solve everything.

The approach we have taken in the rest of the code base is to prefer using the concrete types.  I don't have an issue with revisiting that decision, but for now we should strive to keep the code as consistent as possible.  

In my experience when there is a break in consistency developers assume there was an important reason for breaking with the paradigm that's already in place.  This, in turn, makes it more difficult to do things like refactor and get new developers up to speed.

Reply:
@wilbaker brings up a good point. Let's aim to stay as consistent as we can in the codebase, especially modifying existing code. We can take up coding style changes as separate issues.

Reply:
Moved back to non-interface.